(FACT-2480) Implement Facter::FileUtil for safe read of files

Filipovici-Andrei opened the pull request.

Commit:
(FACT-2480) Added the File Util module

Commit:
Merge branch 'master' into FACT-2480

Commit:
(FACT-2480) Changed combo of File.readable? and File.read/File.readlines with methods from FileHelper module

Commit:
(FACT-2480) Added unit tests for when the required files do not exist

Commit:
Merge branch 'master' into FACT-2480

# Conflicts:
#	spec/facter/resolvers/ssh_resolver_spec.rb

Commit:
(FACT-2480) Fixed rubocop offenses

Commit:
(FACT-2480) Added the Resolvers Utils in the file_loader.rb

Commit:
(FACT-2480) Added missing mocks that were breaking the CI

oanatmaria commented on lib/framework/core/file_loader.rb:
```diff
@@ -37,6 +37,7 @@ def load_lib_dirs(*dirs)
 load_dir(['config'])
 
 load_lib_dirs('resolvers')
+load_lib_dirs('resolvers', 'utils')
```
you can delete 
require "#{ROOT_DIR}/lib/resolvers/utils/fingerprint.rb"
require "#{ROOT_DIR}/lib/resolvers/utils/ssh.rb"
require "#{ROOT_DIR}/lib/resolvers/utils/filesystem_helper.rb"

from the lines above

Filipovici-Andrei replied:
if I remove require "#{ROOT_DIR}/lib/resolvers/utils/filesystem_helper.rb" I get an error saying
uninitialized constant Facter::FilesystemHelper

Commit:
(FACT-2536) Removed unnecessary loaded files

Commit:
(FACT-2480) Fixed some rubocop offenses
Replaced missed file read with open3 with the added safe_read wrapper
Solved file loading issue

IrimieBogdan commented on lib/resolvers/utils/file_helper.rb:
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+module Facter
+  module Resolvers
+    module Utils
+      module FileHelper
```
As this should be use in all instances when we need to read a file, I propose moving it to `/lib/util`

IrimieBogdan replied:
`FileHelper` should be a class with static methods, modules are meant for sharing methods between classes.

Commit:
(FACT-2480) Moved file_helper.rb in facter/util folder

Commit:
(FACT-2480) Changed FileHelper from a module to a class

Commit:
(FACT-2480) Added the FileHelper in file_loader.rb

Commit:
Merge branch 'master' into FACT-2480

# Conflicts:
#	.rubocop_todo.yml

IrimieBogdan commented on lib/resolvers/partitions.rb:
```diff
@@ -41,23 +42,27 @@ def extract_from_device(block_path)
-          populate_partitions("/dev/#{block_path}", block_path) if File.readable?("#{block_path}/loop/backing_file")
-          backing_file = File.read("#{block_path}/loop/backing_file").chomp
-          populate_partitions("/dev/#{block_path}", block_path, backing_file)
+          backing_file = Facter::Util::FileHelper.safe_read("#{block_path}/loop/backing_file").chomp
```
As we are already in the `Facter` module, you can drop `Facter::` from all the places it is used, except tests.

IrimieBogdan commented on lib/resolvers/solaris/filesystems_resolver.rb:
```diff
@@ -15,11 +15,11 @@ def post_resolve(fact_name)
 
           def read_sysdef_file(fact_name)
-            return unless File.readable?('/usr/sbin/sysdef')
+            file_content = Facter::Util::FileHelper.safe_readlines('/usr/sbin/sysdef', nil)
```
This is an executable, reading it as a file will provide incorrect information. You should check if it executable `File.executable?` and than use `Open3.capture3` to execute it.

Commit:
(FACT-2542) Changed how FileHelper is called. Instead of using Facter::Util::FileHelper, now it's Util::FileHelper.
Reverted solaris filesystem fact to using open3 instead of reading the file.
